Add install_python_debian parameter to deb822_repository

pkingstonxyz opened the pull request:
##### SUMMARY

Addresses #85479 by mimicking the functionality of the `install_python_apt` parameter
in the `apt_repository` module by adding a parameter `install_python_debian` to the
`deb822_repository` module. 

Also fixes two broken tests that depended on the previous set of parameters by adding
the new `install_python_debian` parameter to the tests.

##### ISSUE TYPE

- Feature Pull Request
- Test Pull Request

pkingstonxyz committed:
Add python3-debian dependency installation in module

pkingstonxyz committed:
Add tests to exercise automatic dependency installation

pkingstonxyz committed:
Fix broken tests to include new module parameter

pkingstonxyz committed:
Fix dash to be an underscore in module param docs

pkingstonxyz committed:
Add changelog fragment

pkingstonxyz committed:
Add blank lines for pep8 compliance

pkingstonxyz committed:
Fix section name in changelog

pkingstonxyz committed:
Remove trailing whitespace

pkingstonxyz committed:
Add version_added to module documentation

bcoca commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -67,6 +67,17 @@
+              the system Python, set O(install_python_debian=false) and ensure that the Python debian library
+              for your Python version is installed some other way.
+        type: bool
+        default: true
```
default to false, we don't want a repository tool 'auto installing' w/o the user explicitly saying so

pkingstonxyz replied:
Fixed in 91cb8da91b40c00aff010e509da29e1d68136be9 💯

bcoca commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -357,6 +370,21 @@ def write_signed_by_key(module, v, slug):
+def install_python_debian(module, deb_pkg_name):
+
+    if not module.check_mode:
+        apt_get_path = module.get_bin_path('apt-get')
```
at this point we can just use `apt`, apt-get is currently a backwards compatibility shim and might go away

pkingstonxyz replied:
Fixed, thanks for pointing this out. Worth it to open a PR fixing this same issue in the apt_repository module?

bcoca replied:
no, that module is for 'older' debian versions, it was created before `apt` was a thing, this helps distinguish the versions these modules apply to.

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -67,6 +67,17 @@
+            - Runs C(apt-get install python3-debian).
+            - Only works with the system Python. If you are using a Python on the remote that is not
+              the system Python, set O(install_python_debian=false) and ensure that the Python debian library
+              for your Python version is installed some other way.
```
```suggestion
            - Whether to automatically try to install the Python Debian library or not, if it is not already installed.
               Without this library, the module does not work.
            - Runs C(apt-get install python3-debian).
            - Only works with the system Python. If you are using Python on the remote that is not
              the system Python, set V(False) to O(install_python_debian) and ensure that the Python Debian library
              for your Python version is installed in some other way.
```

pkingstonxyz replied:
I made some changes to address some of these suggestions, but on the capitalization of debian: I left it lowercase to point to the python library (ex: `from debian import whatever`) and not to the actual Debian project.

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -357,6 +370,21 @@ def write_signed_by_key(module, v, slug):
+        if apt_get_path:
+            rc, so, se = module.run_command([apt_get_path, 'update'])
+            if rc != 0:
+                module.fail_json(msg="Failed to auto-install %s. Error was: '%s'" % (deb_pkg_name, se.strip()))
```
```suggestion
                module.fail_json(msg=f"Failed update while auto installing {deb_pkg_name} due to '{se.strip()}'")
```

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -357,6 +370,21 @@ def write_signed_by_key(module, v, slug):
+                module.fail_json(msg="Failed to auto-install %s. Error was: '%s'" % (deb_pkg_name, se.strip()))
+            rc, so, se = module.run_command([apt_get_path, 'install', deb_pkg_name, '-y', '-q'])
+            if rc != 0:
+                module.fail_json(msg="Failed to auto-install %s. Error was: '%s'" % (deb_pkg_name, se.strip()))
```
```suggestion
                module.fail_json(msg=f"Failed to auto-install {deb_pkg_name} due to : '{se.strip()}'")
```

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -453,8 +485,54 @@ def main():
+
+        if has_respawned():
+            # this shouldn't be possible; short-circuit early if it happens...
+            module.fail_json(msg="{0} must be installed and visible from {1}.".format(deb_pkg_name, sys.executable))
```
```suggestion
            module.fail_json(msg=f"{deb_pkg_name} must be installed and visible from {sys.executable}.")
```

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -453,8 +485,54 @@ def main():
+        # don't make changes if we're in check_mode
+        if module.check_mode:
+            module.fail_json(msg="%s must be installed to use check mode. "
+                                 "If run normally this module can auto-install it." % deb_pkg_name)
```
```suggestion
            module.fail_json(msg=f"{deb_pkg_name} must be installed to use check mode. "
                                                  "If run normally, this module can auto-install it.")
```

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -453,8 +485,54 @@ def main():
+        if module.params['install_python_debian']:
+            install_python_debian(module, deb_pkg_name)
+        else:
+            module.fail_json(msg='%s is not installed, and install_python_debian is False' % deb_pkg_name)
```
```suggestion
            module.fail_json(msg=f'{deb_pkg_name} is not installed, and install_python_debian is False')
```

Akasurde commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -453,8 +485,54 @@ def main():
+            # this is the end of the line for this process, it will exit here once the respawned module has completed
+        else:
+            # we've done all we can do; just tell the user it's busted and get out
+            module.fail_json(msg=missing_required_lib("python3-debian"),
```
```suggestion
            module.fail_json(msg=missing_required_lib(deb_pkg_name),
```

Akasurde commented on test/integration/targets/deb822_repository/tasks/main.yml:
```diff
@@ -1,13 +1,61 @@
 
+    - name: assert python3-debian already installed
+      assert:
+        that: py3_deb_install.changed == false
```
```suggestion
        that:
           - not py3_deb_install.changed
```

Akasurde commented on test/integration/targets/deb822_repository/tasks/main.yml:
```diff
@@ -1,13 +1,61 @@
+        types: deb
+        uris: "http://example.com/debian/"
+        suites: stable
+        components: main
```
Could you check the success of this task?

pkingstonxyz replied:
An assert could be added for this task's success, but I'm not sure what that would provide. This task is only there because of the `install_python_debian: true`. Its success is measured by the the assertion that `not py3_deb_install.changed`. The actual functionality is already covered by other tests

pkingstonxyz committed:
Change install_python_debian default to false

Also updates tests accordingly.

pkingstonxyz committed:
Move from apt-get to apt

pkingstonxyz committed:
Remove inspect test used to test tests

pkingstonxyz committed:
Fix module docs to reflect apt-get to apt change

pkingstonxyz committed:
Refer more explicitly to the python library debian

Instead of capitalizing debian to Debian, adds C(debian)
to point more explicitly at from debian import ...

pkingstonxyz committed:
Move string formatting to fstrings

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

pkingstonxyz committed:
Merge remote-tracking branch 'origin/add-install-python-apt-to-deb822' into add-install-python-apt-to-deb822

# Conflicts:
#	lib/ansible/modules/deb822_repository.py

pkingstonxyz committed:
Change hard-coded string to var

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

pkingstonxyz committed:
Fix one straggling fstring problem

pkingstonxyz committed:
Update assert to avoid equality check

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

pkingstonxyz committed:
Fix error message and indentation problem.

pkingstonxyz committed:
Set required=True on get_bin_path

Module should fail when apt is (somehow) not installed
on the system.

pkingstonxyz committed:
Fix module to search for debian instead of apt

webknjaz commented on lib/ansible/modules/deb822_repository.py:
```diff
@@ -357,6 +370,21 @@ def write_signed_by_key(module, v, slug):
     return changed, filename, None
 
 
+def install_python_debian(module, deb_pkg_name):
```
Many branches here seem uncovered: https://app.codecov.io/gh/ansible/ansible/commit/a0d56d2f4f32f174bd6c463118a3a38eeb70921b.

Codecov reports some coverage drop of 5.56% for this module. @pkingstonxyz could you see if this could be addressed?

pkingstonxyz replied:
Went ahead and made some in #85586 if you could take a look! My coverage shows 90% with those changes
